### ISSUE TITLE
Explore a 256 and 'true color' API

### ISSUE THREAD

**Issue by jbnicolai** [chalk/chalk#73]
See chalk/supports-color#25 and chalk/ansi-styles#11

Once chalk/supports-color and chalk/ansi-styles land support for 256 and truecolor we might want to expose this in chalk as well.

Any suggestions for a simple, clean API?

**Comment by Qix-** [comment]
The problem I was facing.

```javascript
chalk.bold.red("Hello");
chalk.bold.rgb(255, 0, 0)("Hello"); //-> Not as elegant.
```

**Comment by Qix-** [comment]
Also, another question is: where should everything go?

We have a few moving parts here:

- Color manipulation functionality (hue shifting, lightness adjustments, RGB <-> HSV/HSL conversions, etc.)
- Degredation from truecolor to 256/basic values
- The fact we want to be able to map these to basic colors (i.e. `rgb(255, 0, 0)` should map to `31` instead of `38;2;255;0;0`)
- Listing above combinations (and applying labels to them)
- Styles (bold, italic, underline, inverse, etc.)
- The API for stringing them together

Some of the above are obvious (i.e. *stringing them together* would occur in chalk proper), but the others?

**Comment by Qix-** [comment]
x-post: chalk/ansi-styles#11

**Comment by silverwind** [comment]
Few API suggestions, leaning onto CSS:

````js
chalk.bold.hex('f00')('Hello');
chalk.bold.hex('#f00')('Hello');
chalk.bold['f00']('Hello'); /* probably needs ES6 Proxy */
chalk.bold.hsl(0, 1, .5)('Hello');
chalk.bold.hsl('0', '100%', '50%')('Hello');
chalk.bold.hsl('0, 100%, 50%')('Hello');
chalk.bold.hsl('0 100% 50%')('Hello');
````

**Comment by jbnicolai** [comment]
Hmm, I'm leaning towards the @silverwind's suggestions

**Comment by silverwind** [comment]
I think all three color formats (hex, rgb, hsl) have valid use cases. The double `()` required for rgb and hsl is a bit unfortunate, but I don't see a way around it.

**Comment by jbnicolai** [comment]
I agree. We should probably put exploring ES6 proxies on hold though.

**Comment by sindresorhus** [comment]
ES6 proxies are super slow and will probably stay so for a long time. I don't think we should go with that for now.

**Comment by silverwind** [comment]
Yeah of course. Might be possible to use proxies at a later point, I was just looking for options to shorten the syntax.

**Comment by stevemao** [comment]
What makes `chalk.bold['f00']('Hello');` better than `chalk.bold.hex('#f00')('Hello');`? They both not quite short and one is not better than the other. 

I don't think `chalk.bold.rgb(255, 0, 0)("Hello");` is not as elegant.
https://github.com/jbnicolai/ansi-256-colors syntax is probably even worse.

**Comment by Qix-** [comment]
@stevemao in @jbnicolai 's defense I don't think he was going for having an expressive API. :+1:

**Comment by Qix-** [comment]
Another suggestion (not at all saying it's better):

```javascript
chalk.bold.rgb(45, 81, 99, "some text ").underline("Some other text");
```

**Comment by stevemao** [comment]
I personally just don't think `chalk.bold.rgb(255, 0, 0)("Hello")` is ugly at all. You can very easily alias a colour: `var red = chalk.bold.rgb(255, 0, 0)`.

with `chalk.bold.rgb(45, 81, 99, "some text ")` you need to curry it to achieve the same thing.

**Comment by Qix-** [comment]
@stevemao definitely agreed on that second point.

**Comment by Qix-** [comment]
I don't think we should have separate methods for the type of input. It's pretty easy to figure out what is meant by `.x("#FF0000")`, `.x(255, 0, 0)` and `.x('hsl(0.1, 1.0, 1.0)')`. Instead, perhaps we should consider the output, i.e.: `.basic(...)` (16 colors), `.ansi(...)` (256 colors), and `.true(...)` (16 million colors). I'm sure the names could be improved, but you get what I mean.

**Comment by silverwind** [comment]
I like using a single method. The hsl syntax could probably be improved. Given that the format is number, percentage, percentage we could do a simple check of the second and third values are fractions (`.5`) or strings (`50%`).

**Comment by Qix-** [comment]
Yeah I was thinking so too. Thoughts @sindresorhus and @jbnicolai?

**Comment by Qix-** [comment]
```javascript
var chalk = require('chalk');

console.log(chalk.bold.color('hsl{40, 100, 20}')('Hello, world!'));
```

**Comment by silverwind** [comment]
Not too bad, except I'd probably just seperate with spaces. If you want to hue-rotate for example, it gets a bit ugly though as you have to mess with strings. What about objects?

```js
chalk.bold.color({h: 40, s: 1, l: .2})
chalk.bold.color({r: 200, g: 100, b: 50})
```

**Comment by Qix-** [comment]
That'd work.

**Comment by Qix-** [comment]
Thoughts @sindresorhus / @jbnicolai ?

**Comment by sindresorhus** [comment]
I like @silverwind's suggestion.

**Comment by silverwind** [comment]
On second thought, I think having seperate methods for each format and compatibilty with [color convert](https://github.com/harthur/color-convert) is probably better. The syntax would also be shorter:

```js
chalk.rgb(139, 199, 50)
chalk.hsl(96, 47, 59)
chalk.hsv(96, 50, 78)
chalk.hwb(96, 39, 22)
chalk.cmyk(30, 0, 50, 22)
```
Also with array support:

```js
chalk.rgb([139, 199, 50])
chalk.hsl([96, 47, 59])
chalk.hsv([96, 50, 78])
chalk.hwb([96, 39, 22])
chalk.cmyk([30, 0, 50, 22])
```

**Comment by arthurvr** [comment]
> On second thought, I think having seperate methods for each format and compatibilty with color convert is probably better. The syntax would also be shorter:

:+1:

**Comment by silverwind** [comment]
Also to consider: Clamp values to 0-255 or 0-100 if they're outside the range for their value.

**Comment by jbnicolai** [comment]
I would also strongly prefer separate named methods for each format. Dealing with strings or objects doesn't make dynamic input easier for those using the API.

> Technically, fractions of 1 would be more correct for percentages, but I think 0-100 is easier to use if done consistently.

I'm leaning the other way, but would be happy to be convinced otherwise. Could you explain why you feel 0-100 is easier to use?

> Also to consider: Clamp values to 0-255, 0-100, 00-ff if they're outside the range for their value. Or maybe throwing is better in that case?

Probably throw?

**Comment by silverwind** [comment]
> I'm leaning the other way, but would be happy to be convinced otherwise. Could you explain why you feel 0-100 is easier to use?

Because CSS uses percentages instead of fractions too.

> Probably throw?

Yeah, probably better so user errors are not masked.

**Comment by jbnicolai** [comment]
> Because CSS uses percentages instead of fractions too.

:+1:

**Comment by sindresorhus** [comment]
@bucaran

**Comment by Qix-** [comment]
As well, using 0-100 is one less mathematical operation between chalk and color-convert.

**Comment by mikeerickson** [comment]
Just my 2 cents worth....

I like any solution, so long as it is possible :-)  

Old thread here, any chance this would move forward and adopted in the near future?

**Comment by thomaslindstrom** [comment]
[This looks nice](https://github.com/chalk/chalk/issues/73#issuecomment-124826052). :+1: The added benefit of efficiently being able to turn `chalk.rgb(139, 199, 50)` into a reusable function for later is perfect.

**Comment by Qix-** [comment]
@sindresorhus does that solution work? If so I can throw together a PR.

**Comment by stevemao** [comment]
Why don't you guys release 2.0.0 now and we can always add this in 3.0.0

**Comment by Qix-** [comment]
Or, if @sindresorhus signs off on it we can get it released as 2.0.0 ;) Not enough changes in there right now to really warrant a 2.0.

**Comment by sindresorhus** [comment]
@stevemao There will never be a 3.0.0.

**Comment by rektide** [comment]
[came across mention](https://twitter.com/fanf/status/729825796609740802) of a nicely maintained list of true color capable terminals, which is a handy reference: https://gist.github.com/XVilka/8346728
